In [1]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns

In [2]:
# Suppress Warnings for clean notebook
import warnings
warnings.filterwarnings('ignore')

In [3]:
dataset = pd.read_csv('Melbourne_housing_FULL.csv')
dataset.head()

Suburb             Address  Rooms Type      Price Method SellerG  \
0  Abbotsford       68 Studley St      2    h        NaN     SS  Jellis   
1  Abbotsford        85 Turner St      2    h  1480000.0      S  Biggin   
2  Abbotsford     25 Bloomburg St      2    h  1035000.0      S  Biggin   
3  Abbotsford  18/659 Victoria St      3    u        NaN     VB  Rounds   
4  Abbotsford        5 Charles St      3    h  1465000.0     SP  Biggin   

        Date  Distance  Postcode  ...  Bathroom  Car  Landsize  BuildingArea  \
0  3/09/2016       2.5    3067.0  ...       1.0  1.0     126.0           NaN   
1  3/12/2016       2.5    3067.0  ...       1.0  1.0     202.0           NaN   
2  4/02/2016       2.5    3067.0  ...       1.0  0.0     156.0          79.0   
3  4/02/2016       2.5    3067.0  ...       2.0  1.0       0.0           NaN   
4  4/03/2017       2.5    3067.0  ...       2.0  0.0     134.0         150.0   

   YearBuilt         CouncilArea Lattitude  Longtitude             Regionname  \
0        NaN  Yarra City Council  -37.8014    144.9958  Northern Metropolitan   
1        NaN  Yarra City Council  -37.7996    144.9984  Northern Metropolitan   
2     1900.0  Yarra City Council  -37.8079    144.9934  Northern Metropolitan   
3        NaN  Yarra City Council  -37.8114    145.0116  Northern Metropolitan   
4     1900.0  Yarra City Council  -37.8093    144.9944  Northern Metropolitan   

  Propertycount  
0        4019.0  
1        4019.0  
2        4019.0  
3        4019.0  
4        4019.0  

[5 rows x 21 columns]

In [4]:
# let's use limited columns which makes more sense for serving our purpose
cols_to_use = ['Suburb', 'Rooms', 'Type', 'Method', 'SellerG', 'Regionname', 'Propertycount', 
               'Distance', 'CouncilArea', 'Bedroom2', 'Bathroom', 'Car', 'Landsize', 'BuildingArea', 'Price']
dataset = dataset[cols_to_use]

In [5]:
dataset.head()

Suburb  Rooms Type Method SellerG             Regionname  \
0  Abbotsford      2    h     SS  Jellis  Northern Metropolitan   
1  Abbotsford      2    h      S  Biggin  Northern Metropolitan   
2  Abbotsford      2    h      S  Biggin  Northern Metropolitan   
3  Abbotsford      3    u     VB  Rounds  Northern Metropolitan   
4  Abbotsford      3    h     SP  Biggin  Northern Metropolitan   

   Propertycount  Distance         CouncilArea  Bedroom2  Bathroom  Car  \
0         4019.0       2.5  Yarra City Council       2.0       1.0  1.0   
1         4019.0       2.5  Yarra City Council       2.0       1.0  1.0   
2         4019.0       2.5  Yarra City Council       2.0       1.0  0.0   
3         4019.0       2.5  Yarra City Council       3.0       2.0  1.0   
4         4019.0       2.5  Yarra City Council       3.0       2.0  0.0   

   Landsize  BuildingArea      Price  
0     126.0           NaN        NaN  
1     202.0           NaN  1480000.0  
2     156.0          79.0  1035000.0  
3       0.0           NaN        NaN  
4     134.0         150.0  1465000.0

# Checking for Nan values

In [6]:
dataset.isna().sum()

Suburb               0
Rooms                0
Type                 0
Method               0
SellerG              0
Regionname           3
Propertycount        3
Distance             1
CouncilArea          3
Bedroom2          8217
Bathroom          8226
Car               8728
Landsize         11810
BuildingArea     21115
Price             7610
dtype: int64

In [8]:
# Some feature's missing values can be treated as zero (another class for NA values or absence of that feature)
# like 0 for Propertycount, Bedroom2 will refer to other class of NA values
# like 0 for Car feature will mean that there's no car parking feature with house
cols_to_fill_zero = ['Propertycount', 'Distance', 'Bedroom2', 'Bathroom', 'Car']
dataset[cols_to_fill_zero] = dataset[cols_to_fill_zero].fillna(0)

In [9]:
# other continuous features can be imputed with mean for faster results since our focus is on Reducing overfitting
# using Lasso and Ridge Regression
dataset['Landsize'] = dataset['Landsize'].fillna(dataset.Landsize.mean())
dataset['BuildingArea'] = dataset['BuildingArea'].fillna(dataset.BuildingArea.mean())

In [10]:
dataset

Suburb  Rooms Type Method        SellerG             Regionname  \
0      Abbotsford      2    h     SS         Jellis  Northern Metropolitan   
1      Abbotsford      2    h      S         Biggin  Northern Metropolitan   
2      Abbotsford      2    h      S         Biggin  Northern Metropolitan   
3      Abbotsford      3    u     VB         Rounds  Northern Metropolitan   
4      Abbotsford      3    h     SP         Biggin  Northern Metropolitan   
...           ...    ...  ...    ...            ...                    ...   
34852  Yarraville      4    h     PI            Jas   Western Metropolitan   
34853  Yarraville      2    h     SP        Sweeney   Western Metropolitan   
34854  Yarraville      2    t      S            Jas   Western Metropolitan   
34855  Yarraville      3    h     SP  hockingstuart   Western Metropolitan   
34856  Yarraville      2    h     PI             RW   Western Metropolitan   

       Propertycount  Distance               CouncilArea  Bedroom2  Bathroom  \
0             4019.0       2.5        Yarra City Council       2.0       1.0   
1             4019.0       2.5        Yarra City Council       2.0       1.0   
2             4019.0       2.5        Yarra City Council       2.0       1.0   
3             4019.0       2.5        Yarra City Council       3.0       2.0   
4             4019.0       2.5        Yarra City Council       3.0       2.0   
...              ...       ...                       ...       ...       ...   
34852         6543.0       6.3  Maribyrnong City Council       4.0       1.0   
34853         6543.0       6.3  Maribyrnong City Council       2.0       2.0   
34854         6543.0       6.3  Maribyrnong City Council       2.0       1.0   
34855         6543.0       6.3  Maribyrnong City Council       0.0       0.0   
34856         6543.0       6.3  Maribyrnong City Council       2.0       1.0   

       Car    Landsize  BuildingArea      Price  
0      1.0  126.000000      160.2564        NaN  
1      1.0  202.000000      160.2564  1480000.0  
2      0.0  156.000000       79.0000  1035000.0  
3      1.0    0.000000      160.2564        NaN  
4      0.0  134.000000      150.0000  1465000.0  
...    ...         ...           ...        ...  
34852  3.0  593.000000      160.2564  1480000.0  
34853  1.0   98.000000      104.0000   888000.0  
34854  2.0  220.000000      120.0000   705000.0  
34855  0.0  593.598993      160.2564  1140000.0  
34856  0.0  250.000000      103.0000  1020000.0  

[34857 rows x 15 columns]

# Drop NaN values of Price, since it's our predictive variable we won't impute it

In [11]:
dataset.dropna(inplace=True)
dataset

Suburb  Rooms Type Method        SellerG             Regionname  \
1      Abbotsford      2    h      S         Biggin  Northern Metropolitan   
2      Abbotsford      2    h      S         Biggin  Northern Metropolitan   
4      Abbotsford      3    h     SP         Biggin  Northern Metropolitan   
5      Abbotsford      3    h     PI         Biggin  Northern Metropolitan   
6      Abbotsford      4    h     VB         Nelson  Northern Metropolitan   
...           ...    ...  ...    ...            ...                    ...   
34852  Yarraville      4    h     PI            Jas   Western Metropolitan   
34853  Yarraville      2    h     SP        Sweeney   Western Metropolitan   
34854  Yarraville      2    t      S            Jas   Western Metropolitan   
34855  Yarraville      3    h     SP  hockingstuart   Western Metropolitan   
34856  Yarraville      2    h     PI             RW   Western Metropolitan   

       Propertycount  Distance               CouncilArea  Bedroom2  Bathroom  \
1             4019.0       2.5        Yarra City Council       2.0       1.0   
2             4019.0       2.5        Yarra City Council       2.0       1.0   
4             4019.0       2.5        Yarra City Council       3.0       2.0   
5             4019.0       2.5        Yarra City Council       3.0       2.0   
6             4019.0       2.5        Yarra City Council       3.0       1.0   
...              ...       ...                       ...       ...       ...   
34852         6543.0       6.3  Maribyrnong City Council       4.0       1.0   
34853         6543.0       6.3  Maribyrnong City Council       2.0       2.0   
34854         6543.0       6.3  Maribyrnong City Council       2.0       1.0   
34855         6543.0       6.3  Maribyrnong City Council       0.0       0.0   
34856         6543.0       6.3  Maribyrnong City Council       2.0       1.0   

       Car    Landsize  BuildingArea      Price  
1      1.0  202.000000      160.2564  1480000.0  
2      0.0  156.000000       79.0000  1035000.0  
4      0.0  134.000000      150.0000  1465000.0  
5      1.0   94.000000      160.2564   850000.0  
6      2.0  120.000000      142.0000  1600000.0  
...    ...         ...           ...        ...  
34852  3.0  593.000000      160.2564  1480000.0  
34853  1.0   98.000000      104.0000   888000.0  
34854  2.0  220.000000      120.0000   705000.0  
34855  0.0  593.598993      160.2564  1140000.0  
34856  0.0  250.000000      103.0000  1020000.0  

[27244 rows x 15 columns]

# Let's one hot encode the categorical features

In [12]:
dataset = pd.get_dummies(dataset, drop_first=True)

In [13]:
dataset.head()

Rooms  Propertycount  Distance  Bedroom2  Bathroom  Car  Landsize  \
1      2         4019.0       2.5       2.0       1.0  1.0     202.0   
2      2         4019.0       2.5       2.0       1.0  0.0     156.0   
4      3         4019.0       2.5       3.0       2.0  0.0     134.0   
5      3         4019.0       2.5       3.0       2.0  1.0      94.0   
6      4         4019.0       2.5       3.0       1.0  2.0     120.0   

   BuildingArea      Price  Suburb_Aberfeldie  ...  \
1      160.2564  1480000.0              False  ...   
2       79.0000  1035000.0              False  ...   
4      150.0000  1465000.0              False  ...   
5      160.2564   850000.0              False  ...   
6      142.0000  1600000.0              False  ...   

   CouncilArea_Moorabool Shire Council  CouncilArea_Moreland City Council  \
1                                False                              False   
2                                False                              False   
4                                False                              False   
5                                False                              False   
6                                False                              False   

   CouncilArea_Nillumbik Shire Council  CouncilArea_Port Phillip City Council  \
1                                False                                  False   
2                                False                                  False   
4                                False                                  False   
5                                False                                  False   
6                                False                                  False   

   CouncilArea_Stonnington City Council  CouncilArea_Whitehorse City Council  \
1                                 False                                False   
2                                 False                                False   
4                                 False                                False   
5                                 False                                False   
6                                 False                                False   

   CouncilArea_Whittlesea City Council  CouncilArea_Wyndham City Council  \
1                                False                             False   
2                                False                             False   
4                                False                             False   
5                                False                             False   
6                                False                             False   

   CouncilArea_Yarra City Council  CouncilArea_Yarra Ranges Shire Council  
1                            True                                   False  
2                            True                                   False  
4                            True                                   False  
5                            True                                   False  
6                            True                                   False  

[5 rows x 745 columns]

In [15]:
X = dataset.drop('Price', axis=1)
y = dataset['Price']

In [16]:
from sklearn.model_selection import train_test_split
train_X, test_X, train_y, test_y = train_test_split(X, y, test_size=0.3, random_state=2)

# Let's train our Linear Regression Model on training dataset and check the accuracy on test set

In [17]:
from sklearn.linear_model import LinearRegression
reg = LinearRegression().fit(train_X, train_y)

In [18]:
reg.score(test_X, test_y)


0.13853683161388253

In [19]:
reg.score(train_X, train_y)

0.6827792395792723

**Here training score is 68% but test score is 13.85% which is very low**

**Normal Regression is clearly overfitting the data, let's try other models**

# Using Lasso (L1 Regularized) Regression Model

In [20]:
from sklearn import linear_model
lasso_reg = linear_model.Lasso(alpha=50, max_iter=100, tol=0.1)
lasso_reg.fit(train_X, train_y)

Lasso(alpha=50, max_iter=100, tol=0.1)

In [21]:
lasso_reg.score(test_X, test_y)

0.6636111369404489

In [22]:
lasso_reg.score(train_X, train_y)

0.6766985624766824

# Using Ridge (L2 Regularized) Regression Model

In [23]:
from sklearn.linear_model import Ridge
ridge_reg= Ridge(alpha=50, max_iter=100, tol=0.1)
ridge_reg.fit(train_X, train_y)

Ridge(alpha=50, max_iter=100, tol=0.1)

In [24]:
ridge_reg.score(test_X, test_y)

0.6670848945194958

In [25]:
ridge_reg.score(train_X, train_y)

0.6622376739684328

**We see that Lasso and Ridge Regularizations prove to be beneficial when our Simple Linear Regression Model overfits. These results may not be that contrast but significant in most cases.Also that L1 & L2 Regularizations are used in Neural Networks too**